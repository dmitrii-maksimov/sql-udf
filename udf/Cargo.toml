[package]
name = "udf"
version = "0.4.0-rc.1"
rust-version = "1.65"
edition = "2021"
description = "Easily create user defined functions (UDFs) for MariaDB and MySQL."
repository = "https://github.com/pluots/sql-udf"
readme = "../README.md"
license = "Apache-2.0"
keywords = ["sql", "udf"]
publish = true

[dependencies]
chrono = "0.4.22"
udf-macros = { path = "../udf-macros", version = "0.4.0-rc.1" }
udf-sys = { path = "../udf-sys", version = "0.4.0-rc.1" }

[features]
mock = []

[package.metadata.release]
shared-version = true

[package.metadata.docs.rs]
features = ["mock"]

# Can't run replacements at workspace root. Need to use this "hacky" sort of way.
[[package.metadata.release.pre-release-replacements]]
file = "../CHANGELOG.md"
search = "Unreleased"
replace = "{{version}}"

[[package.metadata.release.pre-release-replacements]]
file = "../CHANGELOG.md"
search = "\\.\\.\\.HEAD"
replace = "...{{tag_name}}"
exactly = 1

[[package.metadata.release.pre-release-replacements]]
file = "../CHANGELOG.md"
search = "ReleaseDate"
replace = "{{date}}"

[[package.metadata.release.pre-release-replacements]]
file = "../CHANGELOG.md"
search = "<!-- next-header -->"
replace = """\
  <!-- next-header -->\n\n\
  ## [Unreleased] - ReleaseDate\n\n\
  ### Added\n\n\
  ### Changed\n\n\
  ### Removed\n\n\
  """
exactly = 1

[[package.metadata.release.pre-release-replacements]]
file = "../CHANGELOG.md"
search = "<!-- next-url -->"
replace = """\
  <!-- next-url -->\n\
  [Unreleased]: https://github.com/pluots/stringmetrics/compare/{{tag_name}}...HEAD\
  """
exactly = 1
