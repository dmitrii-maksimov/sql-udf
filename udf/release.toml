# Replace dependency version in Cargo.toml

[[pre-release-replacements]]
file = "Cargo.toml"
search = 'udf-macros = \{ path = "../udf-macros", version = .* \}'
replace = 'udf-macros = { path = "../udf-macros", version = "{{version}}" }'

[[pre-release-replacements]]
file = "Cargo.toml"
search = 'udf-sys = \{ path = "../udf-sys", version = .* \}'
replace = 'udf-sys = { path = "../udf-sys", version = "{{version}}" }'
