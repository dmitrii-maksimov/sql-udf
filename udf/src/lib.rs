//!
//! Version note: Because of reliance on a feature called GATs, this library
//! requires Rust version >= 1.65 which is currently in beta. If `rustup show`
//! does not show 1.65 or greater under active toolchain, you will need to
//! update:
//!
//! ```sh
//! # nightly can also be used instead of beta
//! rustup default beta
//! rustup update beta
//! ```
//!
//! 1.65 is scheduled to become stable on November 03 2022 UTC, so this message
//! may not be relevant not long after time of writing.
//!
//! # Example
//!
//! Your struct type should hold anything that you want to carry between the
//! functions.
//!
//!
//! ```
//!
//!
//! ```
//!
//! # Behind the Scenes
//!
//! Store the struct to the *ptr before exit
//!
//! Define the basic traits here
//!

#![deny(unsafe_op_in_unsafe_fn)]
#![allow(unused_imports)]
#![allow(unused_variables)]
#![allow(dead_code)]

// Strict clippy
#![warn(
    // clippy::pedantic,
    clippy::cargo,
    // clippy::nursery,
    // clippy::str_to_string,
    // clippy::missing_inline_in_public_items,
    // clippy::exhaustive_enums,
    // clippy::pattern_type_mismatch
)]

extern crate udf_derive;
pub use udf_derive::register;

pub mod ffi;
pub mod prelude;
pub mod traits;
pub mod types;
pub use traits::*;
// Make this inline so we don't show the re-exports
// #[doc(inline)]
pub use types::*;
