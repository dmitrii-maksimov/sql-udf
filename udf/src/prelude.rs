//! Module that can be imported with `use udf::prelude::*;` to quickly get the
//! most often used imports.

pub use std::num::NonZeroU8;

pub use crate::{
    register, AggregateUdf, ArgList, BasicUdf, Init, Process, ProcessError, SqlArg, SqlType, UdfCfg,
};
