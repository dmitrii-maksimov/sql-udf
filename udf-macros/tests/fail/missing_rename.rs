#![allow(unused)]

use udf::prelude::*;

struct MyUdf;

#[register(name = "foo", alias = "bar")]
impl BasicUdf for MyUdf {
    type Returns<'a> = Option<i64>;

    fn init(cfg: &UdfCfg<Init>, args: &ArgList<Init>) -> Result<Self, String> {
        todo!();
    }

    fn process<'a>(
        &'a mut self,
        cfg: &UdfCfg<Process>,
        args: &ArgList<Process>,
        error: Option<NonZeroU8>,
    ) -> Result<Self::Returns<'a>, ProcessError> {
        todo!();
    }
}

#[register]
impl AggregateUdf for MyUdf {
    // Required methods
    fn clear(&mut self, cfg: &UdfCfg<Process>, error: Option<NonZeroU8>) -> Result<(), NonZeroU8> {
        todo!()
    }
    fn add(
        &mut self,
        cfg: &UdfCfg<Process>,
        args: &ArgList<'_, Process>,
        error: Option<NonZeroU8>,
    ) -> Result<(), NonZeroU8> {
        todo!()
    }
}

fn main() {}
