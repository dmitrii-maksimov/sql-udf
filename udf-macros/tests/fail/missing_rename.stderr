error[E0425]: cannot find value `my_udf` in this scope
  --> tests/fail/missing_rename.rs:25:1
   |
25 | #[register]
   | ^^^^^^^^^^^ not found in this scope
   |
   = note: this error originates in the attribute macro `register` (in Nightly builds, run with -Z macro-backtrace for more info)
